제출 형식: ipynb 파일 또는 보고서 파일(pdf, word, hwp)

11. 이번장에는 패션 아이템을 심층 신경망으로 분류하는 프로그램이 있다. 패션 아이템을 기본 MLP로 분류하는 프로그램을 작성하고 심층 신경망과 비교하여 보자. 성능이 얼마나 높아지는가? (10점 만점)


힌트: 해당 문제의 뜻은 예전 예제 코드에서 사용한 기본 하이퍼파리미터들로 사용 시 성능이 어떻게 변화하는지 작성하라는 뜻입니다.

활성화 함수를 모두 sigmoid, 그리고 최적화 기법(optimizier)를 sgd로 바꾸어보세요.

In [76]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow.keras.layers import Input, Dense, Flatten, Dropout
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import SGD
from keras.utils import to_categorical
fashion_mnist = tf.keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

train_images = train_images/ 255.0
test_images = test_images / 255.0

In [97]:
inputs = Input(shape = (28,28))
flatten = Flatten()(inputs)
hidden1 = Dense(128, activation = 'sigmoid')(flatten)
outputs = Dense(10, activation = 'sigmoid')(hidden1)
model = Model(inputs=inputs, outputs=outputs)
model.compile(optimizer='sgd', loss = 'sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size = 128)
test_loss, mlp_test_acc = model.evaluate(test_images, test_labels)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.3799 - loss: 2.0876
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6814 - loss: 1.3682
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.7151 - loss: 1.0792
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.7299 - loss: 0.9333
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.7403 - loss: 0.8463
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7434 - loss: 0.8146


In [98]:
inputs = Input(shape = (28,28))
flatten = Flatten()(inputs)
hidden1 = Dense(128, activation = 'relu')(flatten)
outputs = Dense(10, activation = 'softmax')(hidden1)
model = Model(inputs=inputs, outputs=outputs)
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images, test_labels)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.7763 - loss: 0.6406
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8623 - loss: 0.3832
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8745 - loss: 0.3411
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8838 - loss: 0.3133
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8927 - loss: 0.2919
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8769 - loss: 0.3432


In [101]:
print("mlp : ", mlp_test_acc)
print("dnn : ", test_acc)
print("차이 : ", round(test_acc - mlp_test_acc, 2))

mlp :  0.7412999868392944
dnn :  0.8765000104904175
차이 :  0.14


기존 fashion.py 코드에서 하이퍼파라미터값만 기존mlp코드값으로 바꿨다  
mlp값이 0.14 정도 정확도가 더 낮다.

힌트를 보니 fashion.py코드에서 하이퍼파라미터값만 바꿔서 계산했더니 기존 코드보다 정확도가 낮아짐